README.md
Chores is a simple desktop application to help keep track of recurring chores.  It has some overlap with calendars and to-do lists, but is focused on recurring tasks that aren't appointments (i.e. they need to be completed with a certain approximate frequency, but not necessarily on certain days or at certain times).


debian/changelog
com.github.georgekap1an.chores (1.0) xenial; urgency=low

  * Initial release. 

 -- Jonathan Hebert <jonathan.hebert@icloud.com>  Sun, 08 Oct 2017 13:29:11 -0500


src/main.cpp
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include "mainwindow.h"
#include <gtkmm/application.h>

int main (int argc, char *argv[])
{
  auto app = Gtk::Application::create(argc, argv, "com.github.georgekap1an.chores");

  ChoresMainWindow chores;

  //Shows the window and returns when it is closed.
  return app->run(chores);
}


src/chorelist.h
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include <gtkmm/treemodel.h>
#include <gtkmm/liststore.h>
#include <gtkmm/treemodelfilter.h>
#include <gtkmm/treeview.h>




struct Chore
{
  Glib::ustring name;
  Glib::ustring date;
  int frequency = 0;
  
  Chore(Glib::ustring const& name, Glib::ustring const& date, int frequency);
};

struct ChangeInfo
{
  std::unique_ptr<Chore> before;
  std::unique_ptr<Chore> after;
  
  void clear();
  explicit operator bool();
};


class ChoreColumns : public Gtk::TreeModelColumnRecord
{
public:
  ChoreColumns();

  Gtk::TreeModelColumn<Glib::ustring> name;
  Gtk::TreeModelColumn<Glib::ustring> date;
  Gtk::TreeModelColumn<int> frequency;
};

class ChoreList
{
public:
  ChoreColumns columns_;
  
  ChoreList();
  ~ChoreList();
  
  Gtk::TreeRow get_selected_row();
  void set_search_text(Glib::ustring text);
  void mark_selection_complete();
  void update_selection(Glib::ustring const& name, Glib::ustring const& date, int frequency);
  void delete_selection();
  void add_chore(Glib::ustring const& name, Glib::ustring const& date, int frequency);
  
  void undo();
  bool empty();

  Gtk::TreeView& view();
  
  void save();
  void load();
  
private:
  
  Glib::RefPtr<Gtk::ListStore> list_store_ = Gtk::ListStore::create(columns_);
  Glib::RefPtr<Gtk::TreeModelFilter> list_filter_ = Gtk::TreeModelFilter::create(list_store_);
  Gtk::TreeView tree_view_;
  Glib::ustring search_text_;
  std::string datapath_;
  
  ChangeInfo last_change_;
  
  void select_first_chore();
  bool should_be_visible(Gtk::TreeIter i);
  std::unique_ptr<Chore> chore_from_row(Gtk::TreeRow const& row);
  
};






src/mainwindow.h
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include <gtkmm/menu.h>
#include <glibmm/i18n.h>
#include <gtkmm/window.h>
#include <gtkmm/togglebutton.h>
#include <gtkmm/searchentry.h>
#include <gtkmm/revealer.h>
#include <gtkmm/scrolledwindow.h>
#include <giomm/settings.h>
#include <gtkmm/popover.h>
#include "choredialog.h"
#include "chorelist.h"

class ChoreMenu : public Gtk::Menu
{
public:
  Gtk::MenuItem finished_item;
  Gtk::MenuItem edit_item;
  Gtk::MenuItem remove_item;
  
  ChoreMenu();
};


class ChoresMainWindow : public Gtk::Window
{

public:
  ChoresMainWindow();
  virtual ~ChoresMainWindow();

private:
  void mark_selection_complete();
  void edit_selection();
  void delete_selection();
  void add_chore();
  void undo();

  //Signal handlers:
  void on_list_button_press(GdkEventButton* event);
  bool on_key_press_event(GdkEventKey* key_event) override;
  void on_search_button_click();

  //Member widgets:
  Gtk::HeaderBar header_bar_;
  Gtk::Button close_button_;
  Gtk::Box box_;
  Gtk::Button add_button_;
  Gtk::ToggleButton search_toggle_button_;
  Gtk::SearchEntry search_entry_;
  Gtk::Revealer search_entry_revealer_;
  Gtk::ScrolledWindow scrolled_window_;
  Gtk::Button undo_button_;
  Gtk::Label welcome_message_;
  Gtk::Popover welcome_popover_;
  ChoreList list_;
  ChoreMenu menu_;
  
  // window state
  int current_height_;
  Glib::RefPtr<Gio::Settings> window_height_settings_;
};

src/choredialog.h
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include <gtkmm/button.h>
#include <glibmm/date.h>
#include <gtkmm/dialog.h>
#include <gtkmm/calendar.h>
#include <gtkmm/spinbutton.h>
#include <gtkmm/grid.h>
#include <gtkmm/popover.h>
#include <glibmm/i18n.h>


class ChoreDialog : public Gtk::Dialog
{
public:
  Gtk::Grid grid;
  Gtk::Label name_label;
  Gtk::Entry name_entry;
  Gtk::Label freq_label;
  Gtk::SpinButton freq_entry;
  Gtk::Label date_label;
  Gtk::Calendar date_picker;
  Gtk::Entry date_entry;
  Gtk::Button save_button;
  Gtk::Label popover_message;
  Gtk::Popover popover;
  
  ChoreDialog(Gtk::Window& window);
  void set_edit_mode(Glib::ustring name, int frequency, Glib::Date date);
  Glib::Date get_date();
  void set_date(Glib::Date date);
};

src/mainwindow.cpp
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include "mainwindow.h"

using namespace std;
using namespace Glib;
using namespace Gdk;
using namespace Gtk;
  
ChoreMenu::ChoreMenu() :
    finished_item(_("Mark Finished")),
    edit_item(_("Edit")),
    remove_item(_("Delete"))
{
  append(finished_item);
  append(edit_item);
  append(remove_item);
}




ChoresMainWindow::ChoresMainWindow() :
  welcome_message_(_("Start by Adding a Chore")),
  welcome_popover_(add_button_)
{
  search_entry_.signal_search_changed().connect([this](){
    list_.set_search_text(search_entry_.get_text().casefold());
  });
  search_entry_.set_visible(true);
  search_entry_.set_visibility(true);
  
  search_entry_revealer_.add(search_entry_);
  search_entry_revealer_.set_transition_duration(500);
  search_entry_revealer_.set_transition_type(Gtk::RevealerTransitionType::REVEALER_TRANSITION_TYPE_SLIDE_DOWN);
  search_entry_revealer_.set_reveal_child(false);
  
  list_.view().signal_row_activated().connect([this](const Gtk::TreeModel::Path& path,
                                                const Gtk::TreeViewColumn* col) { edit_selection(); });
  
  list_.view().signal_button_press_event().connect_notify([&](GdkEventButton* event) { on_list_button_press(event); });
  
  
  menu_.finished_item.signal_activate().connect([this](){ mark_selection_complete(); });
  menu_.edit_item.signal_activate().connect([this](){ edit_selection(); });
  menu_.remove_item.signal_activate().connect([this](){ delete_selection(); });
  menu_.show_all(); 
  
  scrolled_window_.set_policy(Gtk::POLICY_NEVER, Gtk::POLICY_AUTOMATIC);
  scrolled_window_.add(list_.view());
  
  add_button_.set_image_from_icon_name("list-add");
  add_button_.signal_clicked().connect([this]() { add_chore(); });
  add_button_.set_tooltip_text(_("Add a New Chore"));
  
  welcome_popover_.add(welcome_message_);
  welcome_popover_.set_border_width(18);
  welcome_popover_.set_modal(false);
  
  
  close_button_.set_image_from_icon_name("window-close-symbolic");
  close_button_.signal_clicked().connect([this](){ close(); });
  
  search_toggle_button_.set_image_from_icon_name("edit-find-symbolic");
  search_toggle_button_.signal_clicked().connect([this](){ on_search_button_click(); });
  search_toggle_button_.set_tooltip_text(_("Find a Chore …")); //need to use real ellipses
  
  undo_button_.set_image_from_icon_name("edit-undo-symbolic");
  undo_button_.set_tooltip_text(_("Undo Last Change"));
  undo_button_.signal_clicked().connect([this]() { undo(); });
  undo_button_.set_sensitive(false);
  
  box_.set_orientation(Gtk::Orientation::ORIENTATION_VERTICAL);
  box_.set_homogeneous(false);
  box_.pack_start(search_entry_revealer_, Gtk::PACK_SHRINK);
  box_.pack_start(scrolled_window_);
  
  header_bar_.set_title(_("Chores"));
  header_bar_.pack_start(close_button_);
  
  header_bar_.pack_end(add_button_);
  header_bar_.pack_end(undo_button_);
  header_bar_.pack_start(search_toggle_button_);
  
  set_title(_("Chores"));
  set_border_width(0);
  
  Gdk::Geometry geometry;
  geometry.min_width = get_width();
  geometry.max_width = get_width();
  geometry.min_height = 400;
  geometry.max_height = 4000;
  
  set_geometry_hints(*this, geometry, Gdk::WindowHints::HINT_MIN_SIZE | Gdk::WindowHints::HINT_MAX_SIZE);
  add(box_);
  
  set_titlebar(header_bar_);
  
  signal_size_allocate().connect([&](Gtk::Allocation& allocation){
    int width;
    get_size(width, current_height_);
  });
  
  window_height_settings_ = Gio::Settings::create("com.github.georgekap1an.chores");
  current_height_ = window_height_settings_->get_int("window-height");
  set_default_size(0, current_height_);
  
  add_events(Gdk::KEY_PRESS_MASK);
  
  this->signal_show().connect([this](){
    if(list_.empty())
    {
      welcome_popover_.show_all();
      search_toggle_button_.set_sensitive(false);
    }
  });
  
  show_all();
}

ChoresMainWindow::~ChoresMainWindow()
{
  window_height_settings_->set_int("window-height", current_height_);
}


void ChoresMainWindow::mark_selection_complete()
{
  list_.mark_selection_complete();
  undo_button_.set_sensitive(true);
}

void ChoresMainWindow::edit_selection()
{
  auto selected_row = list_.get_selected_row();
  Date selected_date;
  selected_date.set_parse(selected_row[list_.columns_.date]);
  ChoreDialog dialog(*this);
  dialog.set_edit_mode(selected_row[list_.columns_.name],
                       selected_row[list_.columns_.frequency],
                       selected_date);
  dialog.show_all();
  
  int response = dialog.run();
  
  if(response == RESPONSE_APPLY)
  {
    list_.update_selection(dialog.name_entry.get_text(),
                           dialog.get_date().format_string("%Y-%m-%d"),
                           dialog.freq_entry.get_value_as_int());
    undo_button_.set_sensitive(true);
  }
}

void ChoresMainWindow::delete_selection()
{
  list_.delete_selection();
  undo_button_.set_sensitive(true);
}

void ChoresMainWindow::add_chore()
{
  if(welcome_popover_.is_visible())
  {
    welcome_popover_.hide();
    search_toggle_button_.set_sensitive(true);
  }
  
  ChoreDialog dialog(*this);
  dialog.show_all();
  int response = dialog.run();
  if(response == RESPONSE_APPLY)
  {
    list_.add_chore(dialog.name_entry.get_text(),
                    dialog.get_date().format_string("%Y-%m-%d"),
                    dialog.freq_entry.get_value_as_int());
    undo_button_.set_sensitive(true);
  }
}

void ChoresMainWindow::undo()
{
  list_.undo();
  undo_button_.set_sensitive(false);
}


void ChoresMainWindow::on_list_button_press(GdkEventButton* event)
{
  if(event->type == GDK_BUTTON_PRESS && event->button == 3)
  {
    menu_.popup(event->button, event->time);
  }
}

bool ChoresMainWindow::on_key_press_event(GdkEventKey* key_event)
{
  if(key_event->keyval == GDK_KEY_Escape)
  {
    if(search_entry_revealer_.get_reveal_child())
    {
      search_entry_revealer_.set_reveal_child(false);
      search_entry_.set_text("");
    }
    return true;
  }
  if(key_event->keyval == GDK_KEY_Delete)
  {
    delete_selection();
    return true;
  }
  if((key_event->state & (CONTROL_MASK | SHIFT_MASK | MOD1_MASK)) == CONTROL_MASK) // ctl ..
  {
    switch(key_event->keyval)
    {
      case GDK_KEY_f: // find
      {
        bool current_state = search_toggle_button_.get_active();
        search_toggle_button_.set_active(!current_state);
        //on_search_button_click();
        return true;
      }
      case GDK_KEY_m:
      {
        mark_selection_complete();
        return true;
      }
      case GDK_KEY_e: // edit
      {
        edit_selection(); // ?
        return true;
      }
      case GDK_KEY_z: // undo
      {
        undo();
        return true;
      }
      case GDK_KEY_n: // new
      {
        add_chore();
        return true;
      }
      case GDK_KEY_q: // quit
      {
        //cout << "ctl-q" << endl;
        close();
        return true;
      }
      case GDK_KEY_w: // close
      {
        //cout << "ctl-w" << endl;
        close();
        return true;
      }
      case GDK_KEY_s: // save
      {
        list_.save();
        return true;
      }
      default:
      {
        break;
      }
    }
  }
  //if the event has not been handled, call the base class
  return Gtk::Window::on_key_press_event(key_event);  
}

void ChoresMainWindow::on_search_button_click()
{
  if(search_entry_revealer_.get_reveal_child())
  {
    search_entry_revealer_.set_reveal_child(false);
    search_entry_.set_text("");
  }
  else
  {
    search_entry_revealer_.set_reveal_child(true);
    search_entry_.grab_focus();
  }
}


  



src/choredialog.cpp
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include "choredialog.h"

using namespace std;
using namespace Glib;
using namespace Gtk;


ChoreDialog::ChoreDialog(Window& window) :
  Dialog(_("Add Chore"), window),
  name_label(_("Name")),
  freq_label(_("Frequency (In Days)")),
  date_label(_("First Due Date")),
  save_button(_("Save")),
  popover_message(_("Enter a Name")),
  popover(name_entry)
{
  name_label.set_halign(Align::ALIGN_END);
  freq_label.set_halign(Align::ALIGN_END);
  date_label.set_halign(Align::ALIGN_END);
  freq_entry.set_range(1, numeric_limits<int>::max());
  freq_entry.set_increments(1,7);
  freq_entry.set_value(7); // default to weekly chore
  
  grid.set_column_spacing(12);
  grid.set_row_spacing(6);
  grid.attach(name_label, 0, 0, 1, 1);
  grid.attach(name_entry, 1, 0, 1, 1);
  grid.attach(freq_label, 0, 1, 1, 1);
  grid.attach(freq_entry, 1, 1, 1, 1);
  grid.attach(date_label, 0, 2, 1, 1);
  grid.attach(date_picker, 1, 2, 1, 1);
  grid.set_border_width(18);
  get_vbox()->add(grid);
  add_button(_("Cancel"), RESPONSE_CANCEL);
  popover.add(popover_message);
  popover.set_border_width(18);
  popover.set_modal(false);
  
  name_entry.signal_changed().connect([this](){
    if(name_entry.get_text_length() > 0 && popover.is_visible())
    {
      popover.hide();
    }
  });
  save_button.signal_clicked().connect([this](){
    if(name_entry.get_text().empty())
    {
      popover.show_all();
      name_entry.grab_focus();
    }
    else
    {
      response(RESPONSE_APPLY);
    }
  });
  this->get_action_area()->pack_end(save_button);
}

void ChoreDialog::set_edit_mode(ustring name, int frequency, Date date)
{
  name_entry.set_text(name);
  freq_entry.set_value(frequency),
  set_date(date);
  set_title(_("Edit Chore"));
  date_label.set_text(_("Next Due Date"));
}

Date ChoreDialog::get_date()
{
  Date date;
  date_picker.get_date(date);
  return date;
}

void ChoreDialog::set_date(Date date)
{
  date_picker.select_month(date.get_month() - 1, date.get_year());
  date_picker.select_day(date.get_day());
}

src/chorelist.cpp
/*
* Copyright (c) 2017 Jonathan Hebert
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public
* License as published by the Free Software Foundation; either
* version 2 of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the
* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA 02110-1301 USA
*
* Authored by: Jonathan Hebert <jonathan.hebert@icloud.com>
*/

#include <iostream>
#include <glibmm/date.h>
#include <glibmm/i18n.h>
#include <glibmm/keyfile.h>
#include <glibmm/miscutils.h>
#include "chorelist.h"

using namespace std;
using namespace Gtk;
using namespace Glib;

Chore::Chore(ustring const& name, ustring const& date, int frequency)
{
  this->name = name;
  this->date = date;
  this->frequency = frequency;
}

void ChangeInfo::clear()
{
  before.reset();
  after.reset();
}

ChangeInfo::operator bool()
{
  return before || after;
}

ChoreColumns::ChoreColumns()
{
  add(name);
  add(date);
  add(frequency);
}

ChoreList::ChoreList() :
  datapath_(build_filename(get_home_dir(), ".chores"))
{
  list_store_->set_sort_column(columns_.date.index(), Gtk::SortType::SORT_ASCENDING);
  list_filter_->set_visible_func([this](Gtk::TreeIter i){ return should_be_visible(i);});
  tree_view_.set_model(list_filter_);
  tree_view_.append_column(_("Chore"), columns_.name);
  tree_view_.append_column(_("Due Date"), columns_.date);
  // built-in search doesn't behave
  // quiet like I want; disable it
  tree_view_.set_search_column(-1);
  load();
}

ChoreList::~ChoreList()
{
  save();
}

TreeRow ChoreList::get_selected_row()
{
  return *(tree_view_.get_selection()->get_selected());
}

void ChoreList::set_search_text(ustring text)
{
  search_text_ = text;
  list_filter_->refilter();
}

void ChoreList::mark_selection_complete()
{
  auto selected_row = get_selected_row();
  
  last_change_.clear();
  last_change_.before = chore_from_row(selected_row);
  
  Date next_due_date;
  next_due_date.set_time_current();
  next_due_date.add_days(selected_row[columns_.frequency]);
  selected_row[columns_.date] = next_due_date.format_string("%Y-%m-%d");
  
  last_change_.after = chore_from_row(selected_row);
  
  select_first_chore();
}

void ChoreList::update_selection(ustring const& name, ustring const& date, int frequency)
{
  auto selected_row = get_selected_row();
  
  last_change_.clear();
  last_change_.before = chore_from_row(selected_row);
  
  selected_row[columns_.name] = name;
  selected_row[columns_.date] = date;
  selected_row[columns_.frequency] = frequency;
  
  last_change_.after = chore_from_row(selected_row);
}

void ChoreList::delete_selection()
{
  auto selected_row = tree_view_.get_selection()->get_selected();
  
  // record change
  last_change_.clear();
  last_change_.before = chore_from_row(*selected_row);
  
  // make change
  auto child_iter = list_filter_->convert_iter_to_child_iter(selected_row);
  list_store_->erase(child_iter);
  
  select_first_chore();
}


void ChoreList::add_chore(ustring const& name, ustring const& date, int frequency)
{
  // make change
  auto new_row = *(list_store_->append());
  new_row[columns_.name] = name;
  new_row[columns_.date] = date;
  new_row[columns_.frequency] = frequency;
  
  // record change
  last_change_.clear();
  last_change_.after = chore_from_row(new_row);
}

void ChoreList::undo()
{
  if(last_change_)
  {
    if(!last_change_.before)
    {
      // undo add be removing
      list_store_->foreach_iter([&](ListStore::iterator const& i) -> bool{
        if(i->get_value(columns_.name) == last_change_.after->name)
        {
          list_store_->erase(i);
          return true;
        }
        return false;
      });
      
    }
    else if(!last_change_.after)
    {
      // undo remove by adding
      auto new_row = *(list_store_->append());
      new_row[columns_.name] = last_change_.before->name;
      new_row[columns_.date] = last_change_.before->date;
      new_row[columns_.frequency] = last_change_.before->frequency;
    }
    else
    {
      // undo edit by restoring original
      for(auto&& row : list_store_->children())
      {
        if(row[columns_.name] == last_change_.after->name)
        {
          row[columns_.name] = last_change_.before->name;
          row[columns_.date] = last_change_.before->date;
          row[columns_.frequency] = last_change_.before->frequency;
          break;
        }
      }
    }
    last_change_.clear();
  }
}

bool ChoreList::empty()
{
  return list_store_->children().empty();
}


TreeView& ChoreList::view()
{
  return tree_view_;
}

void ChoreList::save()
{
  KeyFile keyfile;
  for(auto&& row : list_store_->children())
  {
    keyfile.set_value(row[columns_.name], "due_date", row[columns_.date]);
    keyfile.set_integer(row[columns_.name], "frequency", row[columns_.frequency]);
  }
  bool success = keyfile.save_to_file(datapath_);
  if(!success)
  {
    //TODO use glib logging facility
    cerr << "failed to save data to " << datapath_ << endl;
  }
}

void ChoreList::load()
{
  try
  {
    KeyFile keyfile;
    keyfile.load_from_file(datapath_);
    auto chore_names = keyfile.get_groups();
    for(auto&& chore_name : chore_names)
    {
      auto new_row = *(list_store_->append());
      new_row[columns_.name] = chore_name;
      new_row[columns_.date] = keyfile.get_string(chore_name, "due_date");
      new_row[columns_.frequency] = max(1, keyfile.get_integer(chore_name, "frequency"));
    }
  }
  catch(Exception& e)
  {
    // treat as first use 
  }
}

void ChoreList::select_first_chore()
{
  TreeModel::Path newpath(list_store_->children().begin());
  tree_view_.set_cursor(newpath);
  tree_view_.grab_focus();
}

bool ChoreList::should_be_visible(TreeIter i)
{
  if(search_text_.empty())
  {
    return true;
  }
  else
  {
    auto& row = *i;
    ustring const& name =  row[columns_.name];
    return name.casefold().find(search_text_) !=  ustring::npos;
  }
}

unique_ptr<Chore> ChoreList::chore_from_row(TreeRow const& row)
{
  return make_unique<Chore>(row[columns_.name], row[columns_.date], row[columns_.frequency]);
}







